src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,688 @@
+            }
+
+            return await RewriteSolutionAsync(document,
+                expression, methodCallSites, allOccurrences, parameterName, cancellationToken).ConfigureAwait(false);
+        }
+
+        protected static string GetNewParameterName(SemanticDocument document, TExpressionSyntax expression, CancellationToken cancellationToken)
+        {
+            if (ShouldRemoveVariableDeclaratorContainingExpression(document, expression, out var varDecl))
+            {
+                return varDecl;
+            }
+
+            var semanticFacts = document.Document.GetRequiredLanguageService<ISemanticFactsService>();
+            return semanticFacts.GenerateNameForExpression(document.SemanticModel, expression, capitalize: false, cancellationToken);
+        }
+
+        protected static bool ShouldRemoveVariableDeclaratorContainingExpression(
+            SemanticDocument document, TExpressionSyntax expression, out string varDecl)
+        {
+            var syntaxFacts = document.Document.GetRequiredLanguageService<ISyntaxFactsService>();
+            var expressionDecl = expression.GetRequiredParent().GetRequiredParent();
+            varDecl = "";
+
+            if (syntaxFacts.IsVariableDeclarator(expressionDecl))
+            {
+                varDecl = syntaxFacts.GetIdentifierOfVariableDeclarator(expressionDecl).ValueText;
+                return true;
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Gets the method symbol the expression is enclosed within
+        /// </summary>
+        protected IMethodSymbol? GetMethodSymbolFromExpression(SemanticDocument semanticDocument,
+            TExpressionSyntax expression, CancellationToken cancellationToken)
+        {
+            var methodSymbol = (IMethodSymbol)semanticDocument.SemanticModel.GetRequiredEnclosingSymbol(expression.SpanStart, cancellationToken);
+            var syntaxFacts = semanticDocument.Document.GetRequiredLanguageService<ISyntaxFactsService>();
+            var methodDeclaration = expression.FirstAncestorOrSelf<SyntaxNode>(node => IsContainedInParameterizedDeclaration(node), ascendOutOfTrivia: true);
+            if (methodDeclaration is null)
+            {
+                return null;
+            }
+
+            return (IMethodSymbol)semanticDocument.SemanticModel.GetDeclaredSymbol(methodDeclaration, cancellationToken)!;
+        }
+
+        /// <summary>
+        /// Locates all the call sites of the method that introduced the parameter
+        /// </summary>
+        /// <returns>Dictionary tying all the invocations to each corresponding document</returns>
+        protected static async Task<ImmutableDictionary<Document, List<TInvocationExpressionSyntax>>> FindCallSitesAsync(
+            SemanticDocument document, IMethodSymbol methodSymbol, CancellationToken cancellationToken)
+        {
+            var methodCallSites = new Dictionary<Document, List<TInvocationExpressionSyntax>>();
```

Comment:
don't make a dictionary to convert into a ImmutableDictionary.  Just create and rerturn a MultiMap (or maybe it's called a MultiDictionary).

Reply:
actually, i dont' think MultiDictionary orders the values, so don't use that.  but use an ImmtuableDictionary.Builder instead so you don't have hte intermediary Dictionary.

Reply:
Changed to use ImmutableDictionary.Builder